onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc

```diff
@@ -89,31 +82,59 @@ Status SliceOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
   ORT_RETURN_IF_ERROR(
       SliceOp::PrepareForComputeHelper(input_starts, input_ends, input_axes, input_steps, compute_metadata));
 
-  std::transform(compute_metadata.starts_.cbegin(), compute_metadata.starts_.cend(),
-                 starts.begin(),
-                 [](int64_t i) { return SafeInt<uint32_t>(i); });
-  std::transform(compute_metadata.ends_.cbegin(), compute_metadata.ends_.cend(), compute_metadata.starts_.cbegin(),
-                 sizes.begin(),
-                 [](int64_t i, int64_t j) { return SafeInt<uint32_t>(i - j); });
-  std::transform(compute_metadata.steps_.cbegin(), compute_metadata.steps_.cend(), steps.begin(),
-                 [](int64_t i) { return SafeInt<uint32_t>(i); });
-
-  emscripten::val options = emscripten::val::object();
-  options.set("strides", emscripten::val::array(steps));
-  options.set("label", node.Name());
-  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("slice", inputs,
-                                                                            emscripten::val::array(starts),
-                                                                            emscripten::val::array(sizes),
-                                                                            options);
+  // Check if reverse op is needed.
+  std::vector<uint32_t> reverse_axes;
+  emscripten::val reverse_output = inputs;
+  for (size_t i = 0; i < rank; ++i) {
+    if (compute_metadata.steps_[i] < 0) {
+      reverse_axes.push_back(SafeInt<uint32_t>(i));
+      compute_metadata.steps_[i] = -compute_metadata.steps_[i];
+      compute_metadata.starts_[i] = input_shape[i] - 1 - compute_metadata.starts_[i];
+      compute_metadata.ends_[i] = input_shape[i] - 1 - compute_metadata.ends_[i];
+    }
+  }
+  if (!reverse_axes.empty()) {
+    emscripten::val reverse_options = emscripten::val::object();
+    reverse_options.set("axes", emscripten::val::array(reverse_axes));
+    reverse_options.set("label", node.Name() + "_reverse");
+    reverse_output = model_builder.GetBuilder().call<emscripten::val>("reverse", inputs, reverse_options);
```

Comment:
It'd be better to add data type check for `reverse` and `split` separately in a `HasSupportedInputsImpl`, in case it fails in old versions of Chrome which doesn't support `reverse`. You can do it in a follow-up. :)

Reply:
Got it👌